File: src/test_linked_list.py
"""Tests for Fibonacci, Lucas and sum_series functions."""
import pytest


def test_init():
    """Test if series imports."""
    from linked_list import LinkedList
    test_list = LinkedList()
    assert test_list.head is None


def test_push_method():
    """Test if push method adds new item to linked list."""
    from linked_list import LinkedList
    test_list = LinkedList()
    test_list.push(1)
    assert test_list.head.data == 1


def test_push_method_two():
    """Test if push method adds new item head, retains last head."""
    from linked_list import LinkedList
    test_list = LinkedList()
    test_list.push(1)
    test_list.push(2)
    assert test_list.head.next.data == 1


def test_pop_method_removes_head_value():
    """Tests if pop method returns the value contained in
    the head of the list."""
    from linked_list import LinkedList
    l = LinkedList()
    l.push(1)
    l.pop()
    assert l.head is None


def test_pop_method_returns_head_value():
    """Tests if pop method returns the value contained in
    the head of the list."""
    from linked_list import LinkedList
    l = LinkedList()
    l.push(1)
    assert l.pop() == 1


def test_pop_method_shifts_values_properly():
    """Tests if pop method returns the value contained in
    the head of the list."""
    from linked_list import LinkedList
    l = LinkedList()
    l.push(1)
    l.push(2)
    assert l.pop() == 2
    assert l.head.data == 1


def test_node_has_attribute():
    """Test if new Node has attributes upon initialization."""
    from linked_list import Node
    test_node = Node(5, 5)
    assert hasattr(test_node, 'data')
    assert hasattr(test_node, 'next')


def test_list_has_attribute():
    """Test if new LinkedList has attribute upon initialization."""
    from linked_list import LinkedList
    test_linked_list = LinkedList(5)
    assert hasattr(test_linked_list, 'head')


def test_node_has_given_attribute_values():
    """Test if new Node has given attribute values."""
    from linked_list import Node
    test_node = Node(5, 5)
    assert test_node.data and test_node.next == 5


def test_linked_list_search_returns_none_if_empty():
    """."""
    from linked_list import LinkedList
    l = LinkedList()
    assert l.search(0) is None


def test_linked_list_search_returns_node_val():
    """."""
    from linked_list import LinkedList
    l = LinkedList()
    l.push(1)
    assert l.search(1) == l.head


@pytest.mark.parametrize('n', range(1, 10))
def test_linked_list_search_returns_random_node_val(n):
    """."""
    from linked_list import LinkedList
    from random import randint
    l = LinkedList()
    for i in range(1, n + 1):
        l.push(i)
    search_me = randint(1, n)
    assert l.search(search_me).data == search_me


def test_iterable_upon_initialization():
    """."""
    from linked_list import LinkedList
    a_list = [1, 2, 3]
    l = LinkedList(a_list)
    for item in a_list:
        assert l.search(item).data == item

File: js/test/test-linked-list.js
'use strict';

let linkedList = require('../linked-list');
let chai = require('chai');
let expect = chai.expect;

describe('Functions to test Linked List.', function() {

  it('Tests new node correctly takes in given values and maintains default values.', function() {
    let testNode = new linkedList.Node(5);
    expect(testNode.data).to.equal(5);
    expect(testNode.next).to.equal(null);
  });

  it('Tests linked list can take iterable upon init.', function() {
    let testList = new linkedList.LinkedList([1, 2, 3, 4, 5]);
    expect(testList.size()).to.equal(5);
  })

  it('Test if push method adds new item to linked list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    expect(testList.head.data).to.equal(1);
  });

  it('Test if push method adds new item head, retains last head.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    testList.push(2);
    expect(testList.head.data).to.equal(2);
    expect(testList.head.next.data).to.equal(1);
  });

  it('Test if push increases size of Linked List.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    expect(testList.size()).to.equal(1);
  });

  it('Tests if pop method returns the removed value.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    expect(testList.pop()).to.equal(1);
  });

  it('Tests if pop method resets head if val given is head val.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(2);
    testList.push(1);
    testList.pop();
    expect(testList.head.data).to.equal(2);
  });

  it('Tests if pop method throws error if list is empty.', function() {
    let testList = new linkedList.LinkedList();
    expect(testList.pop).to.throw(Error);
  });

  it('Tests if size method returns correct size of linked list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    testList.push(10);
    testList.push(100);
    expect(testList.size()).to.equal(3);
  });

  it('Tests search method throws error if node containing val not in list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    expect(testList.search).to.throw(Error);
  });

  it('Tests search method returns Node if in linked list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    testList.push(2);
    testList.push(3);
    expect(testList.search(1)).to.be.an.instanceof(linkedList.Node);
  });

  it('Tests remove method throws error if list is empty.', function() {
    let testList = new linkedList.LinkedList();
    expect(testList.remove).to.throw(Error);
  });

  it('Tests remove method throws error if val not in list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1)
    expect(testList.remove).to.throw(Error);
  });

  it('Tests remove method resets head if removing head node.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(15);
    testList.push(30);
    testList.remove(30);
    expect(testList.head.data).to.equal(15);
  });

  it('Tests remove method removes correct node from list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    testList.push(2);
    testList.remove(1);
    expect(testList.head.next).to.be.null;
  });

  it('Tests remove method reduces size of list.', function() {
    let testList = new linkedList.LinkedList();
    testList.push(1);
    testList.push(2);
    testList.remove(1);
    expect(testList.size()).to.equal(1);
  });

  it('Tests display method correctly displays linked list.', function() {
    let testList = new linkedList.LinkedList();
    for (var i = 1; i < 6; i++) {testList.push(i);}
    expect(testList.display()).to.be.string('5, 4, 3, 2, 1');
  });

});
